Fix #5039 : Align policy text and symbols to the left, partial fix for list items

TanishMoral11 opened the pull request:
Fix #5039: 
### PR Description:

This PR fixes issue #5039, where the Terms of Service and Privacy Policy content was incorrectly right-aligned in RTL (Right-to-Left) languages despite being in English. The goal is to keep the English text left-aligned, even when the app is set to an RTL language like Arabic.

### Key Changes:
1. **Updated `PoliciesActivity`**:
   - Changed `PoliciesActivity` to extend `InjectableAutoLocalizedAppCompatActivity`.
   - Added an override for `initializeMixin` to use `ActivityLanguageSource.USE_ENGLISH`, forcing the Policies screen to always use English and ensuring left-to-right alignment.

2. **Modified `InjectableAutoLocalizedAppCompatActivity`**:
   - Replaced the `shouldOnlyUseSystemLanguage = false` boolean with `ActivityLanguageSource.USE_APP_LANGUAGE` for consistency with the new enum-based approach.

3. **Modified `InjectableSystemLocalizedAppCompatActivity`**:
   - Updated to use `ActivityLanguageSource.USE_SYSTEM_LANGUAGE` instead of the old boolean `shouldOnlyUseSystemLanguage = true`.

4. **Enhanced `AppLanguageWatcherMixin`**:
   - Changed the `initialize` method to accept an `ActivityLanguageSource` parameter instead of a boolean.
   - Added support for `USE_ENGLISH` to enforce English localization when specified, alongside `USE_SYSTEM_LANGUAGE` and `USE_APP_LANGUAGE`.

5. **Added `getLocaleFor` in `TranslationController`**:
   - Introduced a new method to fetch the English locale explicitly using `OppiaLanguage.ENGLISH`, supporting the `USE_ENGLISH` option.

6. **Defined `ActivityLanguageSource` Enum in `languages.proto`**:
   - Created an enum with `USE_SYSTEM_LANGUAGE`, `USE_APP_LANGUAGE`, and `USE_ENGLISH` to control activity localization flexibly.

7. **Updated Tests in `AppLanguageWatcherMixinTest`**:
   - Adjusted test cases to use the new `ActivityLanguageSource` enum and added tests for `USE_ENGLISH` to verify English localization.

### Impact:
- The Policies screen now displays English content that stays left-aligned, even in RTL language settings (e.g., Arabic).
- This improves readability and consistency for English text in legal pages, regardless of the app’s language.

## After Changes : 

https://github.com/user-attachments/assets/0112f850-ed6a-46f8-a22c-fb6389bf26da

## Essential Checklist
- [x] The PR title and explanation start with "Fix #5039".
- [x] Changes to any scripts or assets are justified.
- [x] The PR follows the [coding style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] Unnecessary code changes from Android Studio are not included.
- [x] The branch is up-to-date and not named `develop`.
- [x] The PR is assigned to relevant reviewers.

TanishMoral11 committed:
Align policy text and symbols to the left, partial fix for list items

TanishMoral11 commented:
Hey @adhiamboperes ,

I have created a draft PR addressing #5039 , which focuses on ensuring that policy text and symbols (such as bullet points and question marks) are aligned to the left in an LTR format. While this change improves the overall text alignment, there remains an outstanding issue with the alignment of `<li>` (list item) tags when rendered within parsed HTML content.

### Issue Summary:
Although most text and individual symbols are now correctly aligned, list items created using the `<li>` tag within the policy descriptions are not fully aligned to the left. The current span-based approach successfully formats standard text and symbols but does not affect `<li>` elements, leading to inconsistencies in the displayed content.

### Request for Guidance:
I kindly ask for your insights or recommendations on potential solutions for handling the alignment of `<li>` tags in this context. Specifically, any advice on:
- Custom parsing strategies that could be employed for list elements.
- Advanced spans or Android-specific methods to ensure `<li>` tags align correctly within parsed text.

Thanks !!

TanishMoral11 commented:
Hey @adhiamboperes PTAL

adhiamboperes commented:
Thanks @TanishMoral11! The solution for bullets look good. As for the items in <li> tags, One approach attempted by @Vishwajith-Shettigar was to take the raw string, like `val policyDescription` and use regex to extract the reguired tag (in this case `li`) and apply the necessary formatting, then setting the mofidified string to the textview. E.g. in https://github.com/oppia/oppia-android/pull/5286/files and https://github.com/oppia/oppia-android/pull/5212/files.

TanishMoral11 committed:
Fix text alignment and bullet point formatting in PoliciesFragment

TanishMoral11 commented:
Hey @adhiamboperes , PTAL

https://github.com/user-attachments/assets/0124c4ed-7ef6-429d-9d6b-ba9a6cc74c97

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
```diff
@@ -228,43 +224,22 @@ class HtmlParser private constructor(
+        entityType,
+        entityId,
+        imageCenterAlign,
+        consoleLogger,
```
1. Per the KDocs for these two create functions, they were seperated for a reason. I do not think you should combine them without providing sufficient reason.
2. We do not mix named and unamed arguments, and in a situation where there are 3 or more arguments like this, we use named arguments to improve readability. Please revert to named arguments as originally present.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
```diff
@@ -123,17 +125,11 @@ class HtmlParser private constructor(
-      entityType,
-      entityId,
-      imageCenterAlign
+      htmlContentTextView, gcsResourceName, entityType, entityId, imageCenterAlign
```
Nit: Formatting change not required.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
```diff
@@ -123,17 +125,11 @@ class HtmlParser private constructor(
-      htmlContent,
-      imageGetter,
-      computeCustomTagHandlers(supportsConceptCards, htmlContentTextView)
+      htmlContent, imageGetter, computeCustomTagHandlers(supportsConceptCards, htmlContentTextView)
```
Nit: Formatting change not required.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/HtmlParser.kt:
```diff
@@ -289,8 +266,9 @@ class HtmlParser private constructor(
     }
   }
-}
+}
```
Insert new line at EOF.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/LiTagHandler.kt:
```diff
@@ -291,4 +314,4 @@ class LiTagHandler(
       setSpan(mark, length, length, Spanned.SPAN_MARK_MARK)
   }
-}
+}
```
Insert new line at EOF.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/ListItemLeadingMarginSpan.kt:
```diff
@@ -179,4 +181,4 @@ sealed class ListItemLeadingMarginSpan : LeadingMarginSpan {
     override fun getLeadingMargin(first: Boolean) = computedLeadingMargin
   }
-}
+}
```
Insert new line at EOF.

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11.

While the solution appears correct, a number of cleanup changes are required to get the PR merge ready.

Also, the video you have uploaded in the comments does not play. Please re-upload it to the PR description and ensure it can play.

TanishMoral11 committed:
Fix: Remove unnecessary formatting changes and maintain consistent structure

TanishMoral11 commented:
Hey @adhiamboperes PTAL ,

Thank you for your detailed review and valuable feedback. 

I have addressed all your requested adjustments, including:

- **Retaining the separation of the `create` methods** as per your guidance to preserve their distinct purposes.
- **Reverting to named arguments** for improved readability, following your recommendation.
- Ensuring **minor formatting issues** have been resolved, including inserting new lines at EOF where necessary.

Additionally, I have re-uploaded the video demonstration to the PR description, ensuring it is now accessible and plays correctly.

Thank you again for your guidance. I look forward to your review of these updates. Please let me know if there are any further adjustments needed.


https://github.com/user-attachments/assets/73008d0b-b100-4c80-95f4-72fc81339ed8

adhiamboperes commented on app/src/main/java/org/oppia/android/app/policies/PoliciesFragmentPresenter.kt:
```diff
@@ -86,4 +90,4 @@ class PoliciesFragmentPresenter @Inject constructor(
     }
   }
-}
+}
```
Missing newline at EOF.

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11! There is one pending fix, PTAL.

TanishMoral11 committed:
Added Newline At EOF

TanishMoral11 commented:
Hey @adhiamboperes PTAL,

Thank you for pointing that out! I’ve added the missing newline at the end of the file. 
Moving forward, I will ensure that this issue does not occur again.
Thanks!

adhiamboperes commented:
@TanishMoral11, Please check the CI report for failing tests.

TanishMoral11 commented:
Hey @adhiamboperes ,
I reviewed the CI report but am having trouble understanding why the tests are failing. 
Would you be able to provide some guidance or insights on what might be causing this?

adhiamboperes commented:
@TanishMoral11  all the tests are failing with this error. Please check that changes to the MathTagHandler did not conflict with the original usage.
![Screenshot 2024-11-18 at 14 54 24](https://github.com/user-attachments/assets/a60989db-1b00-4c26-a03b-92f24adb1b31)

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Fix left-align policy text

TanishMoral11 committed:
B
Merge branch 'left-align-policy-text-fix' of https://github.com/TanishMoral11/oppia-android into left-align-policy-text-fix

TanishMoral11 committed:
Small Change

TanishMoral11 committed:
Add Old Comments For Future Reference

TanishMoral11 commented:
Hey @adhiamboperes Could You PTAL ,

### Explanation of Implementation

I’ve made updates to handle the left alignment of text, especially bullet points, in the Policies page for both RTL and LTR layouts.

1. **Left-Aligned Bullet Points:**  
   A custom span (`LeftAlignedSymbolsSpan`) was added to ensure that bullet points are always left-aligned, regardless of text direction.

2. **Text Alignment Adjustments:**  
   The text in the policy description was explicitly set to left-aligned for both layout direction (`View.LAYOUT_DIRECTION_LTR`) and text direction (`View.TEXT_DIRECTION_LTR`), ensuring consistent presentation.


Test Passed Screen Shot :
![Screenshot 2024-12-10 211515](https://github.com/user-attachments/assets/8cc398d7-1298-4863-a233-6118c76d6531)


After Changes Video : 
https://github.com/user-attachments/assets/db3a82db-4a71-4aeb-b483-9eb822da7080

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/ListItemLeadingMarginSpan.kt:
```diff
@@ -39,14 +38,13 @@ sealed class ListItemLeadingMarginSpan : LeadingMarginSpan {
-    private val spacingBeforeBullet = resources.getDimensionPixelSize(R.dimen.spacing_before_bullet)
 
     private val bulletDiameter by lazy { bulletRadius * 2 }
+    private val baseMargin = (16f * context.resources.displayMetrics.density).toInt()
```
Here and below, why are we using a hardcoded value of 16f here?

TanishMoral11 replied:
Apologies for the oversight. I’ve replaced the hardcoded value to improve maintainability.

For **Unordered List (UL)**:
``` 
val baseMargin = context.resources.getDimensionPixelSize(R.dimen.spacing_before_bullet)
```
---
For **Ordered List (OL)**:
``` 
val baseMargin = context.resources.getDimensionPixelSize(R.dimen.spacing_before_number_prefix)
```

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/locale/LeftAlignedSymbolsSpan.kt:
```diff
@@ -0,0 +1,40 @@
+import android.graphics.Paint
+import android.text.style.ReplacementSpan
+
+// Custom span to force LTR alignment for symbols as well
+// Custom span to force LTR alignment for all text including symbols
```
This should be a KDOC rather than a comment.

TanishMoral11 replied:
Done.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/ListItemLeadingMarginSpan.kt:
```diff
@@ -153,30 +143,24 @@ sealed class ListItemLeadingMarginSpan : LeadingMarginSpan {
+        // Force left alignment
+        paint.textAlign = Paint.Align.LEFT
+
         val textWidth = Rect().also {
```
Looks like this variable is nolonger used in the new code.

TanishMoral11 replied:
Thanks @adhiamboperes , I’ve removed this unused variable.

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11!

I have some clarifications, and a few comments to solve the CI failures.
1. Please fix the unused attribute error `utility/src/main/java/org/oppia/android/util/parser/html/ListItemLeadingMarginSpan.kt:149:13: warning: variable 'textWidth' is never used`
2.  [LeftAlignedSymbolsSpan.kt](https://github.com/oppia/oppia-android/pull/5573/files#diff-629cfdb57266c5bf09ed4b58981c9004970b425495035e4eef5655f5cef1d374) is a new class, but it does not seem to have been added to a Bazel BUILD file, so Bazel may not be able to find it during build.
3. Is `locale` the correct package for `LeftAlignedSymbolsSpan`? Please verify that that is the same package where similar utilities are created.

Lastly, please do a spot check in a few lessons/revision cards to verify that the bullets behavior remains unaffected.

TanishMoral11 committed:
Remove Ununsed Variable

TanishMoral11 committed:
Remove Ununsed Variable

TanishMoral11 committed:
Resolved All Issues

TanishMoral11 committed:
Resolved All Issues

TanishMoral11 committed:
Improving Formatting

TanishMoral11 commented:
Thanks @adhiamboperes , PTAL !!

> 2\. `[LeftAlignedSymbolsSpan.kt](https://github.com/oppia/oppia-android/pull/5573/files#diff-629cfdb57266c5bf09ed4b58981c9004970b425495035e4eef5655f5cef1d374)` is a new class, but it does not seem to have been added to a Bazel BUILD file, so Bazel may not be able to find it during build.

I Added `LeftAlignedSymbolsSpan.kt` To The Bazel BUILD File But Encountered Test Failures. Could You Help Me Understand Which Tests Are Failing And Any Specific Integration Requirements?

> 3\. Is `locale` the correct package for `LeftAlignedSymbolsSpan`? Please verify that that is the same packake where similar utilities are created.

Yeah, You're Right About The Package Location. I Moved `LeftAlignedSymbolsSpan` To `org.oppia.android.util.parser.html` Since It's An HTML Parsing Utility.

> Lastly, please do a spot check in a few lessons/revision cards to verify that the bullets behavior remains unaffected.

I've Tested The App, And As Yet No Misbehavior Was Detected Related To These Changes.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/BUILD.bazel:
```diff
@@ -46,6 +46,7 @@ kt_android_library(
     name = "list_item_leading_margin_span",
     srcs = [
         "ListItemLeadingMarginSpan.kt",
+        "LeftAlignedSymbolsSpan.kt",
```
@TanishMoral11, to answer your question regarding adding this class to the build file:
1. Sometimes a kt lib can have multiple source files like you have added here, and for `tag_handlers`, but it depends on usecase. E.g., if you want to use `LeftAlignedSymbolsSpan` in a different part of the codebase, you will need to add a dependency on the lib, `list_item_leading_margin_span`, which will provide both `ListItemLeadingMarginSpan` and `LeftAlignedSymbolsSpan`. On the other hand, each class can have it's own lib if they are only used in a standalone way, like most of the other utility classes in this package and their corresponding libs in this BUILD file.

adhiamboperes reviewed (changes requested):
Apologies for taking a while to review this, @TanishMoral11.

A few points to note on Bazel:
1. I have taken a look at your CI failures. The latest error is `app/policies/PoliciesFragmentPresenter_updated.kt:78:56: warning: parameter 'lineIndex' is never used, could be renamed to _
    parsedHtmlDescription.split("\n").forEachIndexed { lineIndex, line ->`. Unless you fix this error, you won't be able to see if there are additional errors. 
2. You can read [this wiki page](https://github.com/oppia/oppia-android/wiki/Interpreting-CI-Results) for information on how to read Bazel CI errors. A youtube video is also linked.
3. You should set aside some time to set up the project build in Bazel, so that you can run Bazel builds locally. Additionally, some more advanced issues you'll encounter will require you to have Bazel set up. The discussions section will be helpful in searching for solutions to some common setup issues.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Removed Unused Variable

TanishMoral11 committed:
Merge branch 'left-align-policy-text-fix' of https://github.com/TanishMoral11/oppia-android into left-align-policy-text-fix

TanishMoral11 commented:
Hey @adhiamboperes , PTAL !!

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/ListItemLeadingMarginSpan.kt:
```diff
@@ -39,14 +38,13 @@ sealed class ListItemLeadingMarginSpan : LeadingMarginSpan {
     private val bulletDiameter by lazy { bulletRadius * 2 }
+    private val baseMargin = context.resources.getDimensionPixelSize((spacing_before_bullet))
+
     private val isRtl by lazy {
```
This variable appears unused.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/ListItemLeadingMarginSpan.kt:
```diff
@@ -123,17 +119,12 @@ sealed class ListItemLeadingMarginSpan : LeadingMarginSpan {
     private val displayLocale: OppiaLocale.DisplayLocale
   ) : ListItemLeadingMarginSpan() {
     private val resources = context.resources
```
This and `displayLocale` above are nolonger used.

TanishMoral11 replied:
Done .

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/parser/html/ListItemLeadingMarginSpan.kt:
```diff
@@ -123,17 +119,12 @@ sealed class ListItemLeadingMarginSpan : LeadingMarginSpan {
-    private val isRtl by lazy {
-      displayLocale.getLayoutDirection() == ViewCompat.LAYOUT_DIRECTION_RTL
-    }
+      2 * longestNumberedItemPrefix.length + baseMargin
```
Even though you have refactored the computation of this, I don't see any current usages.

TanishMoral11 replied:
Done  .

adhiamboperes committed:
Merge branch 'develop' into left-align-policy-text-fix

adhiamboperes reviewed (changes requested):
Thanks @TanishMoral11! I only have comments for one file, PTAL.

I have updated the branch from develop too re-trigger CI so we can see the CI failures, the logs had already expired.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Testing Purpose

TanishMoral11 committed:
Merge branch 'left-align-policy-text-fix' of https://github.com/TanishMoral11/oppia-android into left-align-policy-text-fix

TanishMoral11 committed:
Resolved Issues

TanishMoral11 committed:
Merge branch 'left-align-policy-text-fix' of https://github.com/TanishMoral11/oppia-android into left-align-policy-text-fix

TanishMoral11 commented:
Hey @adhiamboperes , PTAL !!

adhiamboperes commented:
@TanishMoral11, it is not clear to me which changes are in the latest version of the PR. Could you please ensure all checks are passing, or let us know how we can help

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge remote-tracking branch 'upstream/develop' into left-align-policy-text-fix

TanishMoral11 committed:
Resolved Merge Conflict

TanishMoral11 committed:
Reformating

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 commented:
Hey @adhiamboperes , PTAL !!

Thanks.

TanishMoral11 commented:
Hey @adhiamboperes , PTAL !!
This PR is ready for review.
Thanks.

adhiamboperes commented:
![IMG_0041](https://github.com/user-attachments/assets/e351b7f7-6c27-476d-9775-1186bcaacaf9)

@TanishMoral11, did you need help identifying the test failures, or understanding why the tests are failing?

TanishMoral11 commented:
Hi @adhiamboperes , 


> did you need help identifying the test failures, or understanding why the tests are failing?

Yeah , can you please guide me on how to pass all the CI checks? I tried everything and double-checked, but I don’t know why 6 tests failed.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 commented:
Hi @Rd4dev,

In the first approach, I added a `supportLtr` flag to the `HtmlParser` and `LiTagHandler` classes to force left-alignment for the Terms/Privacy text even in RTL. This flag overrides the default RTL alignment for specific text by adjusting the `isRtl` logic. I also updated the bullet/list rendering in `ListItemLeadingMarginSpan` to respect this flag.

I tested manually by running the app in both Arabic (RTL) and English (LTR) languages. All screens worked correctly.

However, I am facing the following issue:

`Expected imageRetriever to be not null.`

![image](https://github.com/user-attachments/assets/3a86af1c-d39f-44cf-bf88-6a49b023e504)

That’s why I changed my approach. In my second approach, I also tested all screens and the complete app in both Arabic (RTL) and English (LTR). This approach works fine for the entire app.

However, by using this method, 6–10 tests are affected. If you think the primary approach is better, no problem, I’ll continue with that approach.

Could you please help me resolve the `Expected imageRetriever to be not null.` issue? I’ve tried everything on my side, but unfortunately, none of the methods worked. I really need some help.

Thanks!

Rd4dev commented:
@TanishMoral11, that's nice to explore different approaches, the reason why I feel the second approach may not be ideal is, it completely removes the RTL language support by ignoring the check for `isRtl` support in `ListItemLeadingMarginSpan` across all existing pages, regardless of locale which shouldn't be the case. This can be visually observed if navigated to `First Test Topic` -> `Revision tab` -> `Test subtopic card`.

| Existing RTL list rendering | Approach2 affecting RTL list render |
|:---:|:---:|
| <img width=250 src="https://github.com/user-attachments/assets/7e6858f7-5826-4286-8c7a-ab4c0100f043"/> | <img width=250 src="https://github.com/user-attachments/assets/bcbfcbaf-9d77-4774-815a-caa23b81cd31"/> |
| <img width=250 src="https://github.com/user-attachments/assets/980d14c0-44f1-40e3-a130-05c8cce5aee1"/> | <img width=250 src="https://github.com/user-attachments/assets/6a75ba7e-da75-474a-9982-e6ddbf843bec" /> |

While the policy and terms page seem to work as expected with the implementation applied, it negatively affects the handling of lists across other pages.

Additionally the `LeftAlignedSymbolsSpan` doesn't seem to have any effect since the parsed text doesn't include ("•") as assumed; the texts are parsed from `<li>`, `<ul>`, and `<ol>` tags.

If possible and haven't yet gone through this blog - https://medium.com/swlh/making-nested-lists-with-android-spannables-in-kotlin-4ad00052912c, do check this out, it explains the essential workings of the functionality for further context.

<h1></h1>

I also tried the PR #5296 implementation locally (which I believe corresponds to Approach 1), and that seems like a valid approach to dynamically support or force LTR rendering for list bulleting and numbering. However, please confirm this with @adhiamboperes.

I ran the tests locally, and they passed fine, but I’m curious about the nature of the test failures. Could you try the first implementation and share what the errors are and if there are any other reasons why approach 1 didn't work? That would be very helpful for understanding the issues.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 commented:
Hi @adhiamboperes , 
Thank you so much for highlighting this PR in today’s meeting. Could you kindly share Ben’s suggestions for resolving this? I’d really appreciate your guidance.

adhiamboperes commented:
@TanishMoral11, I have added the proposed solution to the main issue body, PTAL.

TanishMoral11 committed:
Clean Up

TanishMoral11 committed:
Fix left-align policy text

TanishMoral11 committed:
klint solve

TanishMoral11 committed:
Kdoc Solved

TanishMoral11 commented:
Hi @BenHenning @adhiamboperes , PTAL

Thank you so much for the suggestion! I really love the approach you recommended. I made all the changes, and the implementation is working absolutely fine.

After Changes : 

https://github.com/user-attachments/assets/5258c7a8-338f-4603-b072-b22e4e221231

TanishMoral11 committed:
update tests

adhiamboperes commented on app/src/main/java/org/oppia/android/app/translation/LanguageSource.kt:
```diff
@@ -0,0 +1,24 @@
+ * the user-selected app language, or enforce English as the language for specific activities.
+ */
+enum class LanguageSource {
+  /**
+   * Indicates that the system language (device default) should be used for localization.
+   */
+  SYSTEM_LANGUAGE,
+
+  /**
+   * Indicates that the app-specific language selected by the user should be used for localization.
+   */
```
These look like they could be single line KDocs

TanishMoral11 replied:
Sure ..

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/translation/TranslationController.kt:
```diff
@@ -532,4 +532,10 @@ class TranslationController @Inject constructor(
       }
     }
   }
+  /**
+   * Returns a data provider for the display locale corresponding to the specified language.
+   */
```
Insert newline above. Also, this KDoc looks like it can fit on a single line. Could you please check that?

TanishMoral11 replied:
Done.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/translation/LanguageSource.kt:
I'm thinking about whether this should be a proto enum or a Kotlin enum. Looking at existing implementation patterns, we only use standalone Kotlin enum classes when there are other functions in the class.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/activity/InjectableAutolocalizeEnglishOnlyActivity.kt:
Rename: -> InjectableAutolocalizedEnglishOnlyActivity.kt (Added d to localize to make it an adjective/descriptive)

TanishMoral11 replied:
Thanks @adhiamboperes , for pointing that out. Yeah, it makes more sense now. I have renamed the file and updated all references accordingly, as it needed to be.

adhiamboperes reviewed (changes requested):
Thanks for making these changes @TanishMoral11! I'm happy to see that the solution is working well.

I have left some comments, and please update your PR description and recordings in the main PR body as well.

Do also work on adding tests for the classes that are failing coverage.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Improved

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
move to proto enum

TanishMoral11 committed:
reformatting

TanishMoral11 committed:
Added Test

TanishMoral11 committed:
comments

TanishMoral11 committed:
Added Test

TanishMoral11 commented:
Hi @adhiamboperes , PTAL.

Thanks !!

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Solves regex check

TanishMoral11 committed:
reformatting

TanishMoral11 commented:
Hi @adhiamboperes , PTAL. 
This pr is ready for review.
Thanks!

adhiamboperes commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -34,6 +34,10 @@ test_file_exemption {
   test_file_not_required: true
 }
+test_file_exemption {
+  exempted_file_path: "app/src/main/java/org/oppia/android/app/activity/InjectableAutolocalizedEnglishOnlyActivity.kt"
```
Looks like this activity has been deleted in lates changes ?

TanishMoral11 replied:
Apologies, I just missed it. I have now removed it.

adhiamboperes reviewed (commented):
Thanks @TanishMoral11! 

The changes in the latest commits have been pretty significant compared to the time of my last review.

I am assigning this to @BenHenning for a full pass.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
remove from exemptioons

TanishMoral11 committed:
Merge remote-tracking branch 'origin/left-align-policy-text-fix' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

BenHenning commented on app/src/main/java/org/oppia/android/app/translation/AppLanguageWatcherMixin.kt:
```diff
@@ -39,7 +41,7 @@ class AppLanguageWatcherMixin @Inject constructor(
    *
    * @param shouldOnlyUseSystemLanguage whether only the system language should be used
```
Please update this KDoc now that the method signature has changed.

TanishMoral11 replied:
Done.

BenHenning commented on model/src/main/proto/languages.proto:
```diff
@@ -37,6 +37,32 @@ enum OppiaLanguage {
+
+  // Indicates that the activity should use the system's default language, as configured by the user
+  // in the device's settings. This aligns with the system's locale settings and reflects the user's
+  // preferred language on the device.
```
Suggest adding something like:

```
Note that this may be a necessary mode in cases when an activity cannot load user settings (such as in certain test or early app lifecycle scenarios).
```

(Needs wrapping)

This helps clarify when to use this setting over `USE_APP_LANGUAGE`.

TanishMoral11 replied:
Done.

BenHenning commented on model/src/main/proto/languages.proto:
```diff
@@ -37,6 +37,32 @@ enum OppiaLanguage {
+
+  // Indicates that the activity should use the language selected within the Oppia app. This may
+  // differ from the system's language if the user has chosen a specific language through the app's
+  // settings, providing flexibility for app-specific localization.
```
Suggest adding:

```
Note that this should generally be the default mode used.
```

(May need wrapping)

Since we essentially never want anyone to use the other two unless they are actually necessary.

TanishMoral11 replied:
Done.

BenHenning commented on model/src/main/proto/languages.proto:
```diff
@@ -37,6 +37,32 @@ enum OppiaLanguage {
+// Specifies the source of the language to be used for localizing activities in the Oppia Android app.
+// This enum determines whether an activity should derive its language from the system settings,
+// the app's selected language, or enforce a specific language such as English.
+enum ActivityLanguageSource {
```
Perhaps instead: `ForcedActivityLanguageMode`? This is actually a very 'strong' setting in the sense that it's overriding default Android OS behavior for the activity (in order to make the in-app language selection work correctly).

TanishMoral11 replied:
Yeah, that makes more sense. Thank you for the suggestion!

BenHenning commented on domain/src/main/java/org/oppia/android/domain/translation/TranslationController.kt:
```diff
@@ -532,4 +532,9 @@ class TranslationController @Inject constructor(
+
+  /** Returns a data provider for the display locale corresponding to the specified language. */
+  fun getLocaleFor(language: OppiaLanguage): DataProvider<OppiaLocale.DisplayLocale> {
+    return localeController.retrieveAppStringDisplayLocale(language)
```
Could we perhaps use the `LocaleController` directly for this situation? We have good reason to always go through `TranslationController` but this is a special one-off case, and it seems simpler than expanding the API of `TranslationController` (which would require additional, possibly redundant tests to be added in `TranslationControllerTest`).

TanishMoral11 replied:
Done.

TanishMoral11 replied:
Done.

BenHenning commented on app/src/test/java/org/oppia/android/app/translation/AppLanguageWatcherMixinTest.kt:
```diff
@@ -229,10 +241,21 @@ class AppLanguageWatcherMixinTest {
   }
 
+  @Test
+  fun testMixin_initialized_withEnglish_setsEnglishLocale() {
```
Here & everywhere else in the test: could you maybe force the system locale to be something other than English? Otherwise it's hard to tell whether the `USE_ENGLISH` mode is actually working correctly.

I suggest picking something none of the tests are verifying such as Swahili.

This would lead to a change in the system language test below, as well as the new app language test you added above, but this seems fine since it essentially will demonstrate exactly that the system language is being selected (Swahili) except for this case where 'USE_ENGLISH' forces the system locale to be ignored.

TanishMoral11 replied:
Hi @BenHenning,

I’ve tested setting the system locale to Swahili as you suggested. The `testMixin_initialized_withEnglish_setsEnglishLocale` test passes, confirming that `USE_ENGLISH` correctly overrides the system locale from Swahili to English.

<details>
<summary>Test Details</summary>

![Image](https://github.com/user-attachments/assets/13882d52-17d1-447e-a859-bc0b67f87dba)

</details>

However, switching the system locale to Swahili causes three existing tests to fail:

1. `testMixin_initialized_noAppLanguageChange_doesNothing`: The recreation count is 1 (expected 0).
2. `testMixin_initialized_withAppLanguageChange_sameLanguage_doesNotRecreateActivity`: The recreation count is 1 (expected 0).
3. `testMixin_initialized_withAppLanguageChange_newLanguage_updatesLocale`: The locale remains English (expected Brazilian Portuguese).

These failures seem to stem from how `USE_APP_LANGUAGE` interacts with the system locale change. Given this, should I revert the `DefineAppLanguageLocaleContext` back to English to maintain test stability, or do you recommend another approach to handle these issues?

Thanks!

BenHenning replied:
We definitely should use a non-English system locale here, so please go ahead and make that change.

As for the failing tests, I'll need to see the full stack traces for each of the failures which I can take a look at once CI runs with the follow-up changes.

BenHenning commented on app/src/main/java/org/oppia/android/app/translation/AppLanguageWatcherMixin.kt:
```diff
@@ -68,10 +70,18 @@ class AppLanguageWatcherMixin @Inject constructor(
+      }
+      ActivityLanguageSource.USE_ENGLISH ->
+        translationController.getLocaleFor(OppiaLanguage.ENGLISH)
+      else -> translationController.getSystemLanguageLocale()
```
Probably worth adding a test for this case (the unspecified mode situation).

TanishMoral11 replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/policies/PoliciesActivity.kt:
```diff
@@ -33,6 +35,10 @@ class PoliciesActivity :
     )
   }
 
+  override fun initializeMixin(appLanguageWatcherMixin: AppLanguageWatcherMixin) {
+    appLanguageWatcherMixin.initialize(ActivityLanguageSource.USE_ENGLISH)
+  }
```
Is it possible to test this? Maybe by verifying that the content is in English and left aligned even when the system locale is Arabic?

It would be very nice to ensure this test fails without this initialization override in place, and passes with it.

TanishMoral11 replied:
Done.

BenHenning reviewed (changes requested):
Thanks @TanishMoral11! This is _really_ well done. I have a bunch of comments to clarify or improve testing (mostly because this is a very core and complex system, so it's good to try and be very thorough), but the foundation in the latest PR version is quite good. This seems like a really solid approach.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
clean up

TanishMoral11 committed:
resolved klint

TanishMoral11 committed:
rm from translationcontroller

TanishMoral11 committed:
Added tests

TanishMoral11 commented:
Hi @BenHenning , PTAL.
Thanks!

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 commented:
Hi @BenHenning , PTAL.
Thanks!

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/policies/PoliciesActivityTest.kt:
```diff
@@ -208,6 +209,30 @@ class PoliciesActivityTest {
   }
 
+  @Test
+  fun testPoliciesActivity_withArabicSystemLocale_contentIsInEnglishAndLeftAligned() {
```
This test doesn't actually seem to be setting up an Arabic system locale. Did you verify that this fails correctly with your fix not in place (i.e. reverting the change to `PoliciesActivity` such that it uses the system language instead of forcing English)?

BenHenning commented on app/src/main/java/org/oppia/android/app/translation/AppLanguageWatcherMixin.kt:
```diff
@@ -37,9 +39,12 @@ class AppLanguageWatcherMixin @Inject constructor(
+   *   - `USE_ENGLISH`: Always use English, overriding both system and app language settings.
    */
-  fun initialize(shouldOnlyUseSystemLanguage: Boolean) {
+  fun initialize(languageSource: ForcedActivityLanguageMode) {
```
Nit: should be 'languageMode' per the proto rename. Ditto for the KDoc reference.

TanishMoral11 replied:
Done.

BenHenning reviewed (commented):
Thanks @TanishMoral11! Just a few follow-up comments, though it seems that you may need some help with investigating new failures after changing the default system locale for the mixin tests. I do recommend continuing to try and fix those after you update the PR in case you find a fix before I'm able to look at the CI results next.

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
Merge branch 'develop' into left-align-policy-text-fix

TanishMoral11 committed:
languageMode
